File: src/treemapper/diffctx/postpass.py
from __future__ import annotations

import logging
from collections import defaultdict
from pathlib import Path

from .fragmentation import _create_whole_file_fragment
from .git import CatFileBatch
from .graph import Graph
from .select import SelectionResult, _IntervalIndex
from .types import Fragment, FragmentId

logger = logging.getLogger(__name__)


def _find_dangling_semantic_names(
    selected: list[Fragment],
    graph: Graph,
    frag_by_id: dict[FragmentId, Fragment],
    selected_ids: set[FragmentId],
) -> set[str]:
    dangling: set[str] = set()
    for frag in selected:
        for nbr_id in graph.neighbors(frag.id):
            if nbr_id in selected_ids:
                continue
            if graph.edge_categories.get((frag.id, nbr_id), "") != "semantic":
                continue
            nbr_frag = frag_by_id.get(nbr_id)
            if nbr_frag and nbr_frag.symbol_name:
                dangling.add(nbr_frag.symbol_name.lower())
    return dangling


def _pick_best_fragment(candidates: list[Fragment], selected_ids: set[FragmentId]) -> Fragment | None:
    if any(c.id in selected_ids for c in candidates):
        return None
    sig_candidates = [f for f in candidates if "_signature" in f.kind]
    full_candidates = [f for f in candidates if "_signature" not in f.kind]
    return next(iter(full_candidates or sig_candidates), None)


def _pick_smallest_fitting(
    candidates: list[Fragment],
    selected_ids: set[FragmentId],
    budget_left: int,
) -> Fragment | None:
    ranked = sorted(candidates, key=lambda f: f.token_count)
    for cand in ranked:
        if cand.token_count <= 0 or cand.id in selected_ids:
            continue
        if cand.token_count <= budget_left:
            return cand
    return None


def _coherence_post_pass(
    result: SelectionResult,
    all_fragments: list[Fragment],
    graph: Graph,
    budget: int,
) -> SelectionResult:
    selected_ids = {f.id for f in result.selected}
    interval_idx = _IntervalIndex()
    for f in result.selected:
        interval_idx.add(f.id)
    remaining = budget - result.used_tokens

    name_to_frags: dict[str, list[Fragment]] = {}
    for f in all_fragments:
        if f.symbol_name:
            name_to_frags.setdefault(f.symbol_name.lower(), []).append(f)

    frag_by_id: dict[FragmentId, Fragment] = {f.id: f for f in all_fragments}
    dangling_names = _find_dangling_semantic_names(result.selected, graph, frag_by_id, selected_ids)

    added: list[Fragment] = []
    for name in dangling_names:
        pick = _pick_best_fragment(name_to_frags.get(name, []), selected_ids)
        if pick and pick.token_count <= remaining and pick.id not in selected_ids and not interval_idx.overlaps(pick):
            added.append(pick)
            selected_ids.add(pick.id)
            interval_idx.add(pick.id)
            remaining -= pick.token_count

    if not added:
        return result

    new_selected = result.selected + added
    new_used = result.used_tokens + sum(f.token_count for f in added)
    new_utility = result.utility

    return SelectionResult(
        selected=new_selected,
        reason=result.reason,
        used_tokens=new_used,
        utility=new_utility,
    )


_RESCUE_BUDGET_FRACTION = 0.05
_RESCUE_MIN_SCORE_PERCENTILE = 0.80


def _compute_rescue_threshold(
    all_fragments: list[Fragment],
    rel_scores: dict[FragmentId, float],
    core_ids: set[FragmentId],
) -> float:
    context_scores = sorted(
        (rel_scores.get(f.id, 0.0) for f in all_fragments if f.id not in core_ids and rel_scores.get(f.id, 0.0) > 0),
        reverse=True,
    )
    if not context_scores:
        return float("inf")
    idx = int(len(context_scores) * (1 - _RESCUE_MIN_SCORE_PERCENTILE))
    return context_scores[min(idx, len(context_scores) - 1)]


def _collect_rescue_candidates(
    all_fragments: list[Fragment],
    rel_scores: dict[FragmentId, float],
    core_ids: set[FragmentId],
    selected_ids: set[FragmentId],
    selected_paths: set[Path],
    min_score: float,
    max_tokens: int,
) -> list[Fragment]:
    changed_paths = {fid.path for fid in core_ids}
    candidates = [
        f
        for f in all_fragments
        if f.id not in selected_ids
        and f.id not in core_ids
        and f.path not in changed_paths
        and f.path not in selected_paths
        and rel_scores.get(f.id, 0.0) >= min_score
        and f.token_count <= max_tokens
    ]
    candidates.sort(key=lambda f: rel_scores.get(f.id, 0.0), reverse=True)
    return candidates


def _rescue_nontrivial_context(
    result: SelectionResult,
    all_fragments: list[Fragment],
    rel_scores: dict[FragmentId, float],
    core_ids: set[FragmentId],
    budget: int,
) -> SelectionResult:
    remaining = budget - result.used_tokens
    rescue_budget = min(remaining, int(budget * _RESCUE_BUDGET_FRACTION))
    if rescue_budget <= 0:
        return result

    min_score = _compute_rescue_threshold(all_fragments, rel_scores, core_ids)
    if min_score == float("inf"):
        return result

    selected_ids = {f.id for f in result.selected}
    selected_paths = {f.path for f in result.selected}
    candidates = _collect_rescue_candidates(
        all_fragments, rel_scores, core_ids, selected_ids, selected_paths, min_score, rescue_budget
    )

    interval_idx = _IntervalIndex()
    for f in result.selected:
        interval_idx.add(f.id)

    added: list[Fragment] = []
    used = 0
    for cand in candidates:
        if used + cand.token_count > rescue_budget:
            continue
        if interval_idx.overlaps(cand):
            continue
        added.append(cand)
        interval_idx.add(cand.id)
        used += cand.token_count

    if not added:
        return result

    logger.debug("diffctx: rescued %d nontrivial context fragments (%d tokens)", len(added), used)
    return SelectionResult(
        selected=result.selected + added,
        reason=result.reason,
        used_tokens=result.used_tokens + used,
        utility=result.utility,
    )


def _ensure_changed_files_represented(
    selected: list[Fragment],
    all_fragments: list[Fragment],
    changed_files: list[Path],
    remaining_budget: int,
    root_dir: Path,
    preferred_revs: list[str],
    batch_reader: CatFileBatch | None = None,
) -> list[Fragment]:
    selected_paths = {f.path for f in selected}
    missing_paths = set(changed_files) - selected_paths

    if not missing_paths:
        return selected

    frags_by_path: dict[Path, list[Fragment]] = defaultdict(list)
    for f in all_fragments:
        if f.path in missing_paths:
            frags_by_path[f.path].append(f)

    added: list[Fragment] = []
    budget_left = remaining_budget
    selected_ids = {f.id for f in selected}
    interval_idx = _IntervalIndex()
    for f in selected:
        interval_idx.add(f.id)

    for path in sorted(missing_paths):
        candidates = frags_by_path.get(path, [])
        if not candidates:
            fallback = _create_whole_file_fragment(path, root_dir, preferred_revs, batch_reader)
            candidates = [fallback] if fallback else []

        picked = _pick_smallest_fitting(candidates, selected_ids, budget_left)
        if picked is not None and not interval_idx.overlaps(picked):
            added.append(picked)
            selected_ids.add(picked.id)
            interval_idx.add(picked.id)
            budget_left -= picked.token_count

    if added:
        logger.debug("diffctx: injected %d fragments to cover %d missing changed files", len(added), len(missing_paths))

    return selected + added

File: diffctx/src/postpass.rs
use std::path::{Path, PathBuf};
use std::sync::Arc;

use rustc_hash::{FxHashMap, FxHashSet};

use crate::config::selection::RESCUE;
use crate::fragmentation::create_whole_file_fragment;
use crate::git::CatFileBatch;
use crate::graph::Graph;
use crate::interval::IntervalIndex;
use crate::types::{Fragment, FragmentId};

fn find_dangling_semantic_names(
    selected: &[Fragment],
    graph: &Graph,
    frag_by_id: &FxHashMap<FragmentId, &Fragment>,
    selected_ids: &FxHashSet<FragmentId>,
) -> FxHashSet<String> {
    let mut dangling = FxHashSet::default();
    for frag in selected {
        graph.for_each_forward_neighbor(&frag.id, |nbr_id, _w| {
            if selected_ids.contains(nbr_id) {
                return;
            }
            let cat = graph
                .edge_categories
                .get(&(frag.id.clone(), nbr_id.clone()));
            if cat
                .map(|c| *c != crate::graph::EdgeCategory::Semantic)
                .unwrap_or(true)
            {
                return;
            }
            if let Some(nbr_frag) = frag_by_id.get(nbr_id) {
                if let Some(ref name) = nbr_frag.symbol_name {
                    dangling.insert(name.to_lowercase());
                }
            }
        });
    }
    dangling
}

fn pick_best_fragment<'a>(
    candidates: &[&'a Fragment],
    selected_ids: &FxHashSet<FragmentId>,
) -> Option<&'a Fragment> {
    if candidates.iter().any(|c| selected_ids.contains(&c.id)) {
        return None;
    }
    let full: Vec<&&Fragment> = candidates
        .iter()
        .filter(|f| !f.kind.is_signature())
        .collect();
    let sig: Vec<&&Fragment> = candidates
        .iter()
        .filter(|f| f.kind.is_signature())
        .collect();
    full.first().or(sig.first()).map(|f| **f)
}

fn pick_smallest_fitting(
    candidates: &[Fragment],
    selected_ids: &FxHashSet<FragmentId>,
    budget_left: u32,
) -> Option<Fragment> {
    let mut sorted: Vec<&Fragment> = candidates.iter().collect();
    sorted.sort_by_key(|f| f.token_count);
    for cand in sorted {
        if cand.token_count == 0 || selected_ids.contains(&cand.id) {
            continue;
        }
        if cand.token_count <= budget_left {
            return Some(cand.clone());
        }
    }
    None
}

pub fn coherence_post_pass(
    selected: &mut Vec<Fragment>,
    all_fragments: &[Fragment],
    graph: &Graph,
    budget: u32,
) {
    let selected_ids: FxHashSet<FragmentId> = selected.iter().map(|f| f.id.clone()).collect();
    let mut interval_idx = IntervalIndex::new();
    for f in selected.iter() {
        interval_idx.add(f);
    }
    let used: u32 = selected.iter().map(|f| f.token_count).sum();
    let mut remaining = budget.saturating_sub(used);

    let mut name_to_frags: FxHashMap<String, Vec<&Fragment>> = FxHashMap::default();
    for f in all_fragments {
        if let Some(ref name) = f.symbol_name {
            name_to_frags
                .entry(name.to_lowercase())
                .or_default()
                .push(f);
        }
    }

    let frag_by_id: FxHashMap<FragmentId, &Fragment> =
        all_fragments.iter().map(|f| (f.id.clone(), f)).collect();
    let dangling_names = find_dangling_semantic_names(selected, graph, &frag_by_id, &selected_ids);

    let mut added_ids = selected_ids;
    for name in &dangling_names {
        let candidates = match name_to_frags.get(name) {
            Some(c) => c,
            None => continue,
        };
        let pick = match pick_best_fragment(candidates, &added_ids) {
            Some(p) => p,
            None => continue,
        };
        if pick.token_count <= remaining
            && !added_ids.contains(&pick.id)
            && !interval_idx.overlaps(pick)
        {
            selected.push(pick.clone());
            added_ids.insert(pick.id.clone());
            interval_idx.add(pick);
            remaining = remaining.saturating_sub(pick.token_count);
        }
    }
}

fn compute_rescue_threshold(
    all_fragments: &[Fragment],
    rel_scores: &FxHashMap<FragmentId, f64>,
    core_ids: &FxHashSet<FragmentId>,
) -> f64 {
    let mut context_scores: Vec<f64> = all_fragments
        .iter()
        .filter(|f| !core_ids.contains(&f.id))
        .map(|f| rel_scores.get(&f.id).copied().unwrap_or(0.0))
        .filter(|&s| s > 0.0)
        .collect();
    if context_scores.is_empty() {
        return f64::INFINITY;
    }
    context_scores.sort_by(|a, b| b.partial_cmp(a).unwrap_or(std::cmp::Ordering::Equal));
    let idx = (context_scores.len() as f64 * (1.0 - RESCUE.min_score_percentile)) as usize;
    context_scores[idx.min(context_scores.len() - 1)]
}

pub fn rescue_nontrivial_context(
    selected: &mut Vec<Fragment>,
    all_fragments: &[Fragment],
    rel_scores: &FxHashMap<FragmentId, f64>,
    core_ids: &FxHashSet<FragmentId>,
    budget: u32,
) {
    let used: u32 = selected.iter().map(|f| f.token_count).sum();
    let remaining = budget.saturating_sub(used);
    let rescue_budget = remaining.min((budget as f64 * RESCUE.budget_fraction) as u32);
    if rescue_budget == 0 {
        return;
    }

    let min_score = compute_rescue_threshold(all_fragments, rel_scores, core_ids);
    if min_score == f64::INFINITY {
        return;
    }

    let selected_ids: FxHashSet<FragmentId> = selected.iter().map(|f| f.id.clone()).collect();
    let selected_paths: FxHashSet<Arc<str>> = selected.iter().map(|f| f.id.path.clone()).collect();
    let changed_paths: FxHashSet<Arc<str>> = core_ids.iter().map(|fid| fid.path.clone()).collect();

    let mut candidates: Vec<&Fragment> = all_fragments
        .iter()
        .filter(|f| {
            !selected_ids.contains(&f.id)
                && !core_ids.contains(&f.id)
                && !changed_paths.contains(&f.id.path)
                && !selected_paths.contains(&f.id.path)
                && rel_scores.get(&f.id).copied().unwrap_or(0.0) >= min_score
                && f.token_count <= rescue_budget
        })
        .collect();
    candidates.sort_by(|a, b| {
        let sa = rel_scores.get(&a.id).copied().unwrap_or(0.0);
        let sb = rel_scores.get(&b.id).copied().unwrap_or(0.0);
        sb.partial_cmp(&sa).unwrap_or(std::cmp::Ordering::Equal)
    });

    let mut interval_idx = IntervalIndex::new();
    for f in selected.iter() {
        interval_idx.add(f);
    }

    let mut budget_used = 0u32;
    for cand in candidates {
        if budget_used + cand.token_count > rescue_budget {
            continue;
        }
        if interval_idx.overlaps(cand) {
            continue;
        }
        selected.push(cand.clone());
        interval_idx.add(cand);
        budget_used += cand.token_count;
    }
}

pub fn ensure_changed_files_represented(
    selected: &mut Vec<Fragment>,
    all_fragments: &[Fragment],
    changed_files: &[PathBuf],
    remaining_budget: u32,
    root_dir: &Path,
    preferred_revs: &[String],
    _batch_reader: Option<&mut CatFileBatch>,
) {
    let selected_paths: FxHashSet<String> = selected
        .iter()
        .map(|f| f.id.path.as_ref().to_string())
        .collect();
    let mut missing_paths: Vec<&PathBuf> = changed_files
        .iter()
        .filter(|p| !selected_paths.contains(&p.to_string_lossy().as_ref().to_string()))
        .collect();
    missing_paths.sort();

    if missing_paths.is_empty() {
        return;
    }

    let mut frags_by_path: FxHashMap<String, Vec<Fragment>> = FxHashMap::default();
    for f in all_fragments {
        let path_str = f.id.path.as_ref().to_string();
        if missing_paths
            .iter()
            .any(|p| p.to_string_lossy().as_ref() == path_str)
        {
            frags_by_path.entry(path_str).or_default().push(f.clone());
        }
    }

    let mut budget_left = remaining_budget;
    let mut selected_ids: FxHashSet<FragmentId> = selected.iter().map(|f| f.id.clone()).collect();
    let mut interval_idx = IntervalIndex::new();
    for f in selected.iter() {
        interval_idx.add(f);
    }

    for path in missing_paths.iter().copied() {
        let path_str = path.to_string_lossy().to_string();
        let candidates = frags_by_path.get(&path_str).cloned().unwrap_or_default();
        let candidates = if candidates.is_empty() {
            match create_whole_file_fragment(path, root_dir, preferred_revs, None) {
                Some(f) => vec![f],
                None => continue,
            }
        } else {
            candidates
        };

        if let Some(picked) = pick_smallest_fitting(&candidates, &selected_ids, budget_left) {
            if !interval_idx.overlaps(&picked) {
                budget_left = budget_left.saturating_sub(picked.token_count);
                selected_ids.insert(picked.id.clone());
                interval_idx.add(&picked);
                selected.push(picked);
            }
        }
    }
}
